src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java

```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+            );
+            return false;
+        }
+    }
+
+    //<editor-fold desc="Helper methods">
+    /**
+     * Helper method to check for NaN values in a calculated descriptor result and track their positions.
+     * This method is thread-safe when used with a synchronized LinkedList.
+     *
+     * @param aVector The vector containing calculated descriptor values
+     * @param aStartIndex The start index in the vector for this descriptor
+     * @param numComponents The number of components for this descriptor
+     * @param aMoleculeIndex The index of the current molecule
+     * @param aNanPositions List to track NaN positions (can be null if NaN positions should not be tracked);
+     *                      must be thread-safe for parallel access; will be filled with int[] {[moleculeIndex, componentIndex]}
+     *                      pairs of NaN values
+     * @return true if any NaN values were found, false otherwise
+     */
+    static boolean checkAndTrackNaNValues(float[] aVector, int aStartIndex, int numComponents,
+                                          int aMoleculeIndex, List<int[]> aNanPositions) {
+        boolean foundNaN = false;
+        for (int i = 0; i < numComponents; i++) {
+            if (Float.isNaN(aVector[aStartIndex + i])) {
+                foundNaN = true;
+                // Track NaN position if aNanPositions is provided
+                // Note: aNanPositions must be thread-safe for parallel access
+                if (aNanPositions != null) {
+                    // Synchronize the add operation to ensure thread safety
+                    aNanPositions.add(new int[]{aMoleculeIndex, aStartIndex + i});
+                }
+            }
+        }
+        return foundNaN;
+    }
+
+    /**
+     * Checks if a descriptor is a fingerprint descriptor.
+     *
+     * @param descriptor The descriptor to check
+     * @return true if the descriptor is a fingerprint, false otherwise
+     */
+    static boolean isFingerprint(Descriptor descriptor) {
+        return descriptor == PUBCHEM_FINGERPRINTER ||
+                descriptor == CIRCULAR_FINGERPRINTER_ECFP_0 ||
+                descriptor == CIRCULAR_FINGERPRINTER_FCFP_0 ||
+                descriptor == CIRCULAR_FINGERPRINTER_ECFP_2 ||
+                descriptor == CIRCULAR_FINGERPRINTER_FCFP_2 ||
+                descriptor == CIRCULAR_FINGERPRINTER_ECFP_4 ||
+                descriptor == CIRCULAR_FINGERPRINTER_FCFP_4 ||
+                descriptor == CIRCULAR_FINGERPRINTER_ECFP_6 ||
+                descriptor == CIRCULAR_FINGERPRINTER_FCFP_6 ||
+                descriptor == MACCS_FINGERPRINTER;
+    }
+
+    /**
+     * Converts an IBitFingerprint to a float array.
+     *
+     * @param bitFingerprint The bit fingerprint to convert
+     * @return Float array with 1.0f for set bits and 0.0f for unset bits
+     */
+    private static float[] convertBitFingerprintToFloatArray(IBitFingerprint bitFingerprint) {
```

Comment:
It would be more memory-efficient to not allocate another float array for the fingerprint first but instead work directly on the matrix, like in the other methods.